README.md
# NAME

Kavorka::TraitFor::Parameter::doc - Document method parameters in the signature

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.14+-blue.svg" alt="Requires Perl 5.14+" />
    <img src="https://img.shields.io/badge/coverage-100.0%25-brightgreen.svg" alt="coverage 100.0%" />
    <a href="https://github.com/Csson/p5-Kavorka-TraitFor-Parameter-doc/actions?query=workflow%3Amakefile-test"><img src="https://img.shields.io/github/workflow/status/Csson/p5-Kavorka-TraitFor-Parameter-doc/makefile-test" alt="Build status at Github" /></a>
    </p>
</div>

# VERSION

Version 0.1105, released 2021-02-28.

# SYNOPSIS

    # The class
    use Moops;

    class My::Class using Moose {

        method square(Int $integer does doc('The integer to square.')) {

            return $integer * $integer;

        }

    }

    # Elsewhere
    my $param = My::Class->meta->get_method('square')->signature->params->[1];
    say sprintf '%s %s. %s', $param->optional ? 'Optional' : 'Required',
                             $param->type->name,
                             $param->traits->{'doc'}[0];

    # says 'Required Int. The integer to square.'

# DESCRIPTION

Kavorka::TraitFor::Parameter::doc adds a trait (`doc`) that is useful for documenting in [Moops](https://metacpan.org/pod/Moops) classes created using [Moose](https://metacpan.org/pod/Moose).

# SEE ALSO

- [Kavorka::TraitFor::ReturnType::doc](https://metacpan.org/pod/Kavorka::TraitFor::ReturnType::doc)
- [Moops](https://metacpan.org/pod/Moops)
- [Kavorka](https://metacpan.org/pod/Kavorka)
- [Moose](https://metacpan.org/pod/Moose)

# SOURCE

[https://github.com/Csson/p5-Kavorka-TraitFor-Parameter-doc](https://github.com/Csson/p5-Kavorka-TraitFor-Parameter-doc)

# HOMEPAGE

[https://metacpan.org/release/Kavorka-TraitFor-Parameter-doc](https://metacpan.org/release/Kavorka-TraitFor-Parameter-doc)

# AUTHOR

Erik Carlsson <info@code301.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2021 by Erik Carlsson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Kavorka-TraitFor-Parameter-doc

{{$NEXT}}
 [API Changes]

 [Bug Fixes]

 [Documentation]

 [Enhancements]

 [New Features]

 [Other Changes]

 [Requirements]

 [Stats & Dependency Changes]

 [Testing]

0.1105 2021-02-28T12:03:29Z UTC
 [Other Changes]
 - Replaced Travis with Github Actions
 - Rebuilt with current Dist::Zilla config
 - No functional changes

 [Stats & Dependency Changes]
 - 3 files changed, 53 insertions(+), 2 deletions(-)
 - (dev req) + Badge::Depot 0.0103
 - (dev req) + Badge::Depot::Plugin::Coverage (any)
 - (dev req) + Badge::Depot::Plugin::Githubactions (any)
 - (dev req) + Badge::Depot::Plugin::Perl (any)
 - (dev req) + Dist::Zilla 6.015
 - (dev req) +
   Dist::Zilla::Plugin::Author::CSSON::GithubActions::Workflow::TestWithMakefile
   0.0106
 - (dev req) + Dist::Zilla::Plugin::Authority 1.009
 - (dev req) + Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   0.009
 - (dev req) + Dist::Zilla::Plugin::ChangeStats::Dependencies::Git 0.0200
 - (dev req) + Dist::Zilla::Plugin::Git 2.046
 - (dev req) + Dist::Zilla::Plugin::Git::CheckFor::CorrectBranch 0.014
 - (dev req) + Dist::Zilla::Plugin::Git::Commit (any)
 - (dev req) + Dist::Zilla::Plugin::MetaConfig (any)
 - (dev req) + Dist::Zilla::Plugin::NextRelease::Grouped 0.0200
 - (dev req) + Dist::Zilla::Plugin::PromptIfStale 0.057
 - (dev req) + Dist::Zilla::Plugin::RewriteVersion::Transitional 0.007
 - (dev req) + Dist::Zilla::Plugin::Test::EOF 0.0501
 - (dev req) + Dist::Zilla::Plugin::Test::EOL 0.18
 - (dev req) + Dist::Zilla::Plugin::Test::NoTabs 0.15
 - (dev req) + Dist::Zilla::Plugin::Test::Version 1.09
 - (dev req) + Pod::Weaver::Section::GenerateSection 1.01
 - (dev req) + Test::EOF (any)
 - (dev req) + Test::EOL (any)
 - (dev req) + Test::More 0.88
 - (dev req) + Test::NoTabs 1.4
 - (dev req) + Test::Version 1
 - (dev req) ~ Dist::Zilla::Plugin::CheckChangesHasContent (any) --> 0.011
 - (dev req) ~ Dist::Zilla::Plugin::Clean (any) --> 0.07
 - (dev req) ~ Dist::Zilla::Plugin::CopyFilesFromBuild (any) --> 0.170880
 - (dev req) ~ Dist::Zilla::Plugin::Git::Contributors (any) --> 0.035
 - (dev req) ~ Dist::Zilla::Plugin::GithubMeta (any) --> 0.54
 - (dev req) ~ Dist::Zilla::Plugin::InstallRelease (any) --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::MetaProvides::Package (any) -->
   2.004003
 - (dev req) ~ Dist::Zilla::Plugin::PodWeaver (any) --> 4.006
 - (dev req) ~ Dist::Zilla::Plugin::ReadmeAnyFromPod (any) --> 0.163250
 - (dev req) ~ Dist::Zilla::Plugin::RunExtraTests (any) --> 0.029
 - (dev req) ~ Dist::Zilla::Plugin::Test::Compile (any) --> 2.058
 - (dev req) ~ Dist::Zilla::Plugin::Test::ReportPrereqs (any) --> 0.027
 - (dev req) ~ Pod::Weaver::Section::Badges (any) --> 0.0401
 - (dev req) ~ Test::Warnings (any) --> 0.026
 - (dev req) - Dist::Iller::Config::Author::CSSON
 - (dev req) - Dist::Zilla::Plugin::MetaProvides::Class
 - (dev req) - Dist::Zilla::Plugin::NextRelease
 - (dev req) - Dist::Zilla::Plugin::NextVersion::Semantic
 - (dev req) - Dist::Zilla::Plugin::OurPkgVersion
 - (dev req) - Dist::Zilla::Plugin::PodnameFromClassname
 - (dev req) - Dist::Zilla::Plugin::Prereqs::Plugins
 - (dev req) - Dist::Zilla::Plugin::PreviousVersion::Changelog
 - (dev req) - Dist::Zilla::Plugin::ReversionOnRelease
 - (dev req) - Dist::Zilla::Plugin::TravisYML
 - (dev req) - Pod::Weaver::Section::Homepage::DefaultCPAN
 - (dev req) - Pod::Weaver::Section::Source::DefaultGitHub
 - (dev sug) + Dist::Iller 0.1411
 - (dev sug) + Dist::Iller::Config::Author::CSSON 0.0328

0.1104 2016-01-26T19:10:56Z UTC
 [Other Changes]
 - Rebuilt with current Dist::Iller::Config::Author::CSSON.

 [Requirements]
 - Moops dependency moved to testing where it belongs.

 [Testing]
 - Modify test class to avoid a potential kwalitee issue.

0.1103 2015-02-16T20:32:18Z CET
 [Meta]
 - Added travis config.

 [Requirements]
 - Specified versions on some prereqs.
 - Updated dist.ini/weaver.ini (a few new plugins).

0.1102 2015-01-17T03:48:47Z CET
 [Meta]
 - Added Moose as prereq.

0.1101 2015-01-15T16:11:06Z CET
 [Documentation]
 - Fixed documentation for ::ReturnType::doc.

 [Enhancements]
 - Added test for ::ReturnType::doc.

0.1100 2015-01-14T00:53:54Z CET
 [New Features]
 - Added ReturnType::doc

0.1000 2015-01-12T00:27:46Z CET
 - Initial release


lib/Kavorka/TraitFor/Parameter/Test.pm
use Moops;

#PODNAME: Kavorka::TraitFor::Parameter::Test

class Kavorka::TraitFor::Parameter::Test using Moose {

    has an_attr => (
        is => 'ro',
        isa => Int|Str,
        traits => ['Documented'],
        required => 0,
        default => 'helloo',
        documentation => 'Use it if you want to',
    );
    

    method square(Int $integer does doc('The integer to square.')) {
        return $integer * $integer;
    }
}

__END__

=pod

:pinter :classname Kavorka::TraitFor::Parameter::Test


=head1 SYNOPSIS

    # The class
    use Moops;

    class My::Class using Moose {

        method tester(Int $integer does doc('This does something important.')) {
            return $integer;
        }

    }

    # Elsewhere
    my $param = My::Class->meta->get_method('do_something')->signature->params->[1];
    say sprintf '%s %s. %s', $param->optional ? 'Optional' : 'Required',
                             $param->type->name,
                             $param->traits->{'doc'}[0];

    # says 'Required Int. This does something important.'

=head1 DESCRIPTION

Kavorka::TraitFor::Parameter::doc adds a trait (C<doc>) that is useful for documenting in L<Moops> classes created using L<Moose>.

=head1 ATTRIBUTES

:pinter :attributes


=head1 SEE ALSO

=for :list
* L<Moops>
* L<Kavorka>
* L<Moose>

=cut


t/corpus/lib/MyKavorkaParamTraitTest.pm
use Moops;

class MyKavorkaParamTraitTest using Moose {

    method square(Int $integer does doc('The integer to square.') --> Int does doc('The integer squared.')) {
        return $integer * $integer;
    }
}


lib/Kavorka/TraitFor/ReturnType/doc.pm
use 5.14.0;
use strict;
use warnings;

package Kavorka::TraitFor::ReturnType::doc;

our $VERSION = '0.1106';
# ABSTRACT: Document return types in the signature
# AUTHORITY

use Moo::Role;

1;

__END__

=pod

=head1 SYNOPSIS

    # The class
    use Moops;

    class My::Class using Moose {

        method square(Int $integer --> Int does doc('The squared integer.')) {

            return $integer * $integer;

        }

    }

    # Elsewhere
    my $return_type = My::Class->meta->get_method('square')->signature->return_types->[0];
    say sprintf 'Returns an %s. %s', $return_type->type->name, $return_type->traits->{'doc'}[0];

    # says 'Returns an Int. The squared integer.'

=head1 DESCRIPTION

Kavorka::TraitFor::ReturnType::doc adds a trait (C<doc>) that is useful for documenting in L<Moops> classes created using L<Moose>.

=head1 SEE ALSO

=for :list
* L<Kavorka::TraitFor::Parameter::doc>
* L<Moops>
* L<Kavorka>
* L<Moose>

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.017.
use strict;
use warnings;

use 5.014000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Document method parameters in the signature",
  "AUTHOR" => "Erik Carlsson <info\@code301.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Kavorka-TraitFor-Parameter-doc",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.014000",
  "NAME" => "Kavorka::TraitFor::Parameter::doc",
  "PREREQ_PM" => {
    "Module::Runtime" => "0.014",
    "Moo::Role" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Moops" => "0.034",
    "Moose" => "2.0000",
    "Test::More" => "0.96"
  },
  "VERSION" => "0.1106",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Module::Runtime" => "0.014",
  "Moo::Role" => 0,
  "Moops" => "0.034",
  "Moose" => "2.0000",
  "Test::More" => "0.96"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
